internal/dcache/rpc/server/handler.go

```diff
@@ -1878,6 +1879,241 @@ refreshFromClustermapAndRetry:
+	}
+
+	wg.Wait()
+
+	rpcResp.Responses[rvInfo.rvName] = &models.PutChunkResponseOrError{
+		Response: resp,
+		Error:    rpcErr,
+	}
+
+	log.Debug("ChunkServiceHandler::PutChunkEx: Received response for %s/%s (file id %s, offset in MiB %v): %v",
+		rvInfo.rvName, req.Request.Chunk.Address.MvName, req.Request.Chunk.Address.FileID,
+		req.Request.Chunk.Address.OffsetInMiB, rpc.PutChunkExResponseToString(rpcResp))
+
+	return rpcResp, nil
+}
+
+// This method forwards the PutChunk request to the next RVs in the list.
+// It calls the PutChunkEx to the first RV in the list passing the next RVs to it.
+func (h *ChunkServiceHandler) forwardPutChunk(ctx context.Context, req *models.PutChunkRequest, rvs []*models.RVNameAndState) *models.PutChunkExResponse {
+	common.Assert(req != nil)
+	common.Assert(req.Chunk != nil)
+	common.Assert(req.Chunk.Address != nil)
+
+	log.Debug("ChunkServiceHandler::forwardPutChunk: Forwarding PutChunk to RVs: %v, request: %v",
+		rpc.ComponentRVsToString(rvs), rpc.PutChunkRequestToString(req))
+
+	//
+	// No more RVs to forward the PutChunk request to, so return empty response.
+	//
+	if len(rvs) == 0 {
+		return &models.PutChunkExResponse{
+			Responses: make(map[string]*models.PutChunkResponseOrError),
+		}
+	}
+
+	currRV := rvs[0]
+	common.Assert(currRV != nil)
+	common.Assert(cm.IsValidRVName(currRV.Name), currRV.Name)
+	common.Assert(cm.IsValidComponentRVState(dcache.StateEnum(currRV.State)), currRV.State)
+
+	var nextRVs []*models.RVNameAndState
+	if len(rvs) > 1 {
+		nextRVs = rvs[1:]
+	}
+
+	var rpcResp *models.PutChunkExResponse
+	var err error
+
+	//
+	// Omit writing to RVs in “offline” or “outofsync” state. It’s ok to omit them as the chunks not
+	// written to them will be copied to them when the mv is (soon) resynced.
+	// Otoh if an RV is in “syncing” state then any new chunk written to it may not be copied by the
+	// ongoing resync operation as the source RV may have been already gone past the enumeration stage
+	// and hence won’t consider this chunk for resync, and hence those MUST have the chunks mandatorily
+	// copied to them.
+	//
+	// For outofsync or offline RVs, add nil PutChunkResponseOrError to the responses map so that
+	// the client can know that these RVs were skipped and the chunk was not written to them.
+	//
+	if currRV.State == string(dcache.StateOffline) ||
+		currRV.State == string(dcache.StateOutOfSync) {
+		log.Debug("ChunkServiceHandler::forwardPutChunk: Skipping %s/%s (RV state: %s, MV state: %s)",
+			currRV.Name, req.Chunk.Address.MvName, currRV.State)
+
+		rpcResp = h.forwardPutChunk(ctx, req, nextRVs)
+		common.Assert(rpcResp != nil)
+		common.Assert(rpcResp.Responses != nil) // rpcResp.Responses should not be nil
+
+		rpcResp.Responses[currRV.Name] = nil // nil PutChunkResponseOrError for skipped RVs
+
+	} else if currRV.State == string(dcache.StateOnline) || currRV.State == string(dcache.StateSyncing) {
+
```

Comment:
This checks must be done only at the originator, once the originator decides who all to fwd to, from that point all that the intermediate nodes do is - write to their local RV and blindly forward to the next RV (with properly updated nextRVs list). On the response side, it splices the status of its local PutChunk() and the slice response it got from the next RV it fwded to and relays back the response to its sender. So eventually the originator will get a sliced response, for each RV.